From bb40b72c6ba1c6db2218b8cdb2d58f37d31551f9 Mon Sep 17 00:00:00 2001
From: KyleWong <kang.wang1988@gmail.com>
Date: Sun, 23 Dec 2018 16:29:05 +0800
Subject: [PATCH] Add intel emulation layer logic for arm.

---
 crypto/cpu-arm-linux.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/crypto/cpu-arm-linux.c b/crypto/cpu-arm-linux.c
index 839b632b0..18a1d747d 100644
--- a/crypto/cpu-arm-linux.c
+++ b/crypto/cpu-arm-linux.c
@@ -276,6 +276,12 @@ static unsigned long get_hwcap2_cpuinfo(const STRING_PIECE *cpuinfo) {
   return ret;
 }
 
+// is_intel returns true if /proc/cpuinfo reports that it is for an intel CPU.
+static int is_intel(const STRING_PIECE *cpuinfo) {
+  return cpuinfo_field_equals(cpuinfo, "Hardware", "placeholder");
+}
+
+
 // has_broken_neon returns one if |in| matches a CPU known to have a broken
 // NEON unit. See https://crbug.com/341598.
 static int has_broken_neon(const STRING_PIECE *cpuinfo) {
@@ -300,6 +306,16 @@ void OPENSSL_cpuid_setup(void) {
   cpuinfo.data = cpuinfo_data;
   cpuinfo.len = cpuinfo_len;
 
+  // If we are in a broken emulation environment, and the CPU reports that it is
+  // not an ARM CPU, then we should not trust the values reported by
+  // getauxval(), etc., and so bail out here.
+  if (is_intel(&cpuinfo)) {
+    g_has_broken_neon = 1;
+    OPENSSL_free(cpuinfo_data);
+    return;
+  }
+
+
   // |getauxval| is not available on Android until API level 20. If it is
   // unavailable, read from /proc/self/auxv as a fallback. This is unreadable
   // on some versions of Android, so further fall back to /proc/cpuinfo.
-- 
2.17.2 (Apple Git-113)

